gadopt/gplates.py

```diff
@@ -0,0 +1,449 @@
+        """
+        Creates a pygplates.GPML velocity feature from specified coordinates.
+
+        This function takes a set of coordinates and converts them into a GPML velocity feature,
+        suitable for use with pyGplates. The created feature is a collection of points on a sphere,
+        each representing a node in a velocity mesh, used for calculating tectonic plate velocities
+        at those points.
+
+        Args:
+            coords (numpy.ndarray): An array of coordinates (shape [# of points, 3]) representing
+                                    points on a sphere where velocities are to be calculated.
+
+        Returns:
+            pygplates.FeatureCollection: A feature collection containing the velocity mesh nodes.
+        """
+
+        # Add points to a multipoint geometry
+        multi_point = pygplates.MultiPointOnSphere(
+            [pygplates.PointOnSphere(x=coords[i, 0], y=coords[i, 1], z=coords[i, 2], normalise=True)
+             for i in range(np.shape(coords)[0])]
+        )
+
+        # Create a feature containing the multipoint feature, and defined as MeshNode type
+        meshnode_feature = pygplates.Feature(pygplates.FeatureType.create_from_qualified_string('gpml:MeshNode'))
+        meshnode_feature.set_geometry(multi_point)
+        meshnode_feature.set_name('Velocity Mesh Nodes from pygplates')
+
+        output_feature_collection = pygplates.FeatureCollection(meshnode_feature)
+
+        return output_feature_collection
+
+    def _interpolate_seeds_u(self, target_coords):
+        """
+        Interpolates seed velocities onto target coordinates.
+
+        This method generates a KD-tree of seed points with numerical values, then uses
+        this tree to find the nearest neighbors of the target coordinates. It interpolates
+        the velocities of these neighbors to the target points, applying a weighted average
+        based on distance. If any target point is exceptionally close to a seed point,
+        it directly assigns the seed's velocity to avoid division by zero errors.
+
+        Args:
+            target_coords (numpy.ndarray): Array of target coordinates for velocity interpolation.
+
+        Returns:
+            numpy.ndarray: Interpolated velocities at the target coordinates.
+        """
+        # calculate velocities here
+        seeds_u = self._calc_velocities(
+            velocity_domain_features=self.velocity_domain_features,
+            topology_features=self.topology_features,
+            rotation_model=self.rotation_model,
+            time=round(self.reconstruction_time, 0),
+            delta_time=self.delta_time)
+
+        seeds_u = np.array([i.to_xyz() for i in seeds_u]) *\
+            ((1e-2 * pyGplatesConnector.velocity_non_dim_factor) / (self.scaling_factor * pyGplatesConnector.yrs2sec))
+
+        # generate a KD-tree of the seeds points that have a numerical value
+        tree = cKDTree(data=self.seeds[seeds_u[:, 0] == seeds_u[:, 0], :], leafsize=16)
```

Comment:
I don't understand the point of doing `seeds_u[:,0] == seeds_u[:,0]`, doesn't this just give an array of `True` that's as long as the first dimension? You end up using this a few times in the indexing too, maybe factor it out?

Any reason to specify `leafsize=16`? This is the default anyway, but maybe let scipy decide that?

Reply:
When pygplates calculates the surface velocities, it should find the associated plate polygon for each point. For some geologic ages, the produced files include polygons that do not entirely cover the Erath's surface. These inevitably result in an empty plate id, which implies that no surface velocity can be calculated by gplates.

Reply:
In the absense of any solution to deal with these points, we do interpolation between the all that actually has a solution. Here, instead of building arrays of coordinates and the velocities that are found for those coordinates, I keep the coordinates the same and assign Nan for points that don't have values. 
Then for building the tree, I only pass in seeds with `seeds_u[:,0] == seeds_u[:,0]`, which are essentially non Nan points.

Reply:
It is probably more readable to use `numpy.isfinite` or `numpy.isnan`, if you do not mind.

Reply:
Agreet with @Patol75 here. Please update.

Reply:
Agreed! Thanks @Patol75